src/devices/cpu/drcbearm64.h

```diff
@@ -0,0 +1,286 @@
+
+	virtual void reset() override;
+	virtual int execute(uml::code_handle &entry) override;
+	virtual void generate(drcuml_block &block, const uml::instruction *instlist, uint32_t numinst) override;
+	virtual bool hash_exists(uint32_t mode, uint32_t pc) override;
+	virtual void get_info(drcbe_info &info) override;
+	virtual bool logging() const override { return false; }
+
+private:
+	class be_parameter
+	{
+		static int const REG_MAX = 30;
+
+	public:
+		// parameter types
+		enum be_parameter_type
+		{
+			PTYPE_NONE = 0,        // invalid
+			PTYPE_IMMEDIATE,       // immediate; value = sign-extended to 64 bits
+			PTYPE_INT_REGISTER,    // integer register; value = 0-REG_MAX
+			PTYPE_FLOAT_REGISTER,  // floating point register; value = 0-REG_MAX
+			PTYPE_MEMORY,          // memory; value = pointer to memory
+			PTYPE_MAX
+		};
+
+		typedef uint64_t be_parameter_value;
+
+		be_parameter() : m_type(PTYPE_NONE), m_value(0) { }
+		be_parameter(const be_parameter &param) : m_type(param.m_type), m_value(param.m_value) { }
+		be_parameter(uint64_t val) : m_type(PTYPE_IMMEDIATE), m_value(val) { }
+		be_parameter(drcbe_arm64 &drcbe, const uml::parameter &param, uint32_t allowed);
+
+		static inline be_parameter make_ireg(int regnum) { assert(regnum >= 0 && regnum < REG_MAX); return be_parameter(PTYPE_INT_REGISTER, regnum); }
+		static inline be_parameter make_freg(int regnum) { assert(regnum >= 0 && regnum < REG_MAX); return be_parameter(PTYPE_FLOAT_REGISTER, regnum); }
+		static inline be_parameter make_memory(void *base) { return be_parameter(PTYPE_MEMORY, reinterpret_cast<be_parameter_value>(base)); }
+		static inline be_parameter make_memory(const void *base) { return be_parameter(PTYPE_MEMORY, reinterpret_cast<be_parameter_value>(const_cast<void *>(base))); }
+
+		bool operator==(const be_parameter &rhs) const { return (m_type == rhs.m_type && m_value == rhs.m_value); }
+		bool operator!=(const be_parameter &rhs) const { return (m_type != rhs.m_type || m_value != rhs.m_value); }
+
+		be_parameter_type type() const { return m_type; }
+		uint64_t immediate() const { assert(m_type == PTYPE_IMMEDIATE); return m_value; }
+		uint32_t ireg() const { assert(m_type == PTYPE_INT_REGISTER); assert(m_value < REG_MAX); return m_value; }
+		uint32_t freg() const { assert(m_type == PTYPE_FLOAT_REGISTER); assert(m_value < REG_MAX); return m_value; }
+		void *memory() const { assert(m_type == PTYPE_MEMORY); return reinterpret_cast<void *>(m_value); }
+
+		bool is_immediate() const { return (m_type == PTYPE_IMMEDIATE); }
+		bool is_int_register() const { return (m_type == PTYPE_INT_REGISTER); }
+		bool is_float_register() const { return (m_type == PTYPE_FLOAT_REGISTER); }
+		bool is_memory() const { return (m_type == PTYPE_MEMORY); }
+
+		bool is_immediate_value(uint64_t value) const { return (m_type == PTYPE_IMMEDIATE && m_value == value); }
+
+		be_parameter(be_parameter_type type, be_parameter_value value) : m_type(type), m_value(value) { }
```

Comment:
Could you please group this logically with the other constructors?

Reply:
This was meant to be private but I deleted that part somehow. I can still move it up if you think it'd be better.